bindgen/codegen/serialize.rs

```diff
@@ -0,0 +1,234 @@
+use std::fmt::{self, Write};
+
+use crate::callbacks::IntKind;
+use crate::DEFAULT_NON_EXTERN_FNS_SUFFIX;
+
+use crate::ir::context::{BindgenContext, TypeId};
+use crate::ir::function::{Function, FunctionKind};
+use crate::ir::ty::{FloatKind, TypeKind};
+
+#[derive(Debug)]
+pub(crate) enum Error {
+    Serialize(String),
+    Fmt(fmt::Error),
+}
+
+impl From<fmt::Error> for Error {
+    fn from(err: fmt::Error) -> Self {
+        Self::Fmt(err)
+    }
+}
+
+impl From<String> for Error {
+    fn from(err: String) -> Self {
+        Self::Serialize(err)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct CItem {
+    header: String,
+    code: String,
+}
+
+impl CItem {
+    pub(crate) fn from_function(
+        function: &Function,
+        ctx: &BindgenContext,
+    ) -> Result<Self, Error> {
+        match function.kind() {
+            FunctionKind::Function => {
+                let signature_type = ctx.resolve_type(function.signature());
+                match signature_type.kind() {
```

Comment:
Rather than nesting this so much let's do:

```
if function.kind() != FunctionKind::Function {
    return Err(...);
}
let signature_type = ctx.resolve_type(..);
let signature = match signature_type.kind() {
    TypeKind::Function(signature) => signature,
    _ => unreachable!(),
};
let mut buf = ...
// all the other stuff.